source/common/http/http2/codec_impl.cc

```diff
@@ -409,10 +409,9 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
           ASSERT(!nghttp2_session_check_server_session(session_));
           stream->waiting_for_non_informational_headers_ = false;
 
-          // This can only happen in the client case in a response, when we received a 1xx to
-          // start out with. In this case, raise as headers. nghttp2 message checking guarantees
-          // proper flow here.
-          ASSERT(!stream->headers_->Status() || stream->headers_->Status()->value() != "100");
+          // Even if we have :status 100 in the client case in a response, when
+          // we received a 1xx to start out with, nghttp2 message checking
+          // guarantees proper flow here.
           stream->decodeHeaders();
```

Comment:
Agree if this can happen, the assert is no good.  When you say nghttp2 guarantees proper flow, does it?  I believe it's legal to have 100, 100, 200 and I don't think our code handles that at all - not sure if it'd result in a double decode headers or what.  Do we have a regression test for that to make sure  it doesn't choke in some weird way?

Reply:
I was assuming you had already established this with the previous comment; if not, what was meant here by "nghttp2 message checking guarantees proper flow here"?

Reply:
That's Matt's comment not mine :-)
https://github.com/envoyproxy/envoy/pull/562/files

I don't think we have a test case for trying to send 100, 100, 200.  Given it's allowed by the spec I am not sure what would happen.  Looking at code, think the 200 would be handled as trailers and I have no idea what would happen if there were actually trailers but I was hoping we could fuzz that as well (or just 100, 100, 100 200) to make sure it's nothing bad.

Reply:
@alyssawilk finally got back to looking at this. https://github.com/envoyproxy/envoy/blob/c4211b3195c885a1547070f02acdea0eb88e316a/test/common/http/http2/codec_impl_test.cc#L191 seems to be the test you're after; it does a double `:100` and you end up with a stream reset. I'll merge this based on that, LMK if there is more to it than this from your thinking.